apps/backend/src/services/boundary-extraction-service.ts

```diff
@@ -0,0 +1,160 @@
+import type { Pool, PoolClient } from "pg"
+import { withTransaction } from "../db"
+import { ConversationRepository, type Conversation } from "../repositories/conversation-repository"
+import { MessageRepository, type Message } from "../repositories/message-repository"
+import { StreamRepository } from "../repositories/stream-repository"
+import { OutboxRepository } from "../repositories/outbox-repository"
+import type { BoundaryExtractor, ExtractionContext, ConversationSummary } from "../lib/boundary-extraction/types"
+import { addStalenessFields } from "../lib/conversation-staleness"
+import { conversationId } from "../lib/id"
+import { ConversationStatuses, StreamTypes } from "@threa/types"
+import { logger } from "../lib/logger"
+
+const WINDOW_SIZE = 5
+
+export class BoundaryExtractionService {
+  constructor(
+    private pool: Pool,
+    private extractor: BoundaryExtractor
+  ) {}
+
+  async processMessage(messageId: string, streamId: string, workspaceId: string): Promise<Conversation | null> {
+    return withTransaction(this.pool, async (client) => {
+      const message = await MessageRepository.findById(client, messageId)
+      if (!message) {
+        logger.warn({ messageId }, "Message not found for boundary extraction")
+        return null
+      }
+
+      const stream = await StreamRepository.findById(client, streamId)
+      if (!stream) {
+        logger.warn({ streamId }, "Stream not found for boundary extraction")
+        return null
+      }
+
+      const recentMessages = await MessageRepository.list(client, streamId, { limit: WINDOW_SIZE })
```

Comment:
I think we should look for messages surrounding `messageId`, not just recent. Hyperactive channels and/or slow processing will yield incorrect results.